packages/tests/shadydom/event-path.html

```diff
@@ -930,6 +937,123 @@
+        const pushToEventLog = (e) => {
+          eventLog.push({currentTarget: e.currentTarget, eventPhase: e.eventPhase});
+        };
+
+        // `ShadyDOM.wrapIfNeeded(outerNoSlot).shadowRoot` will also see this event in its path,
+        // but event listeners on shadow roots are instead redirected by Shady
+        // DOM to their host, causing the phase to incorrect (AT_TARGET).
+        ShadyDOM.wrap(outerNoSlot).addEventListener('some-event', pushToEventLog);
+        ShadyDOM.wrap(outerNoSlot).addEventListener('some-event', pushToEventLog, true);
+        ShadyDOM.wrap(outerDiv).addEventListener('some-event', pushToEventLog);
+        ShadyDOM.wrap(outerDiv).addEventListener('some-event', pushToEventLog, true);
+        ShadyDOM.wrap(innerNoSlot).addEventListener('some-event', pushToEventLog);
+        ShadyDOM.wrap(innerNoSlot).addEventListener('some-event', pushToEventLog, true);
+        ShadyDOM.wrap(target).addEventListener('some-event', pushToEventLog);
+        ShadyDOM.wrap(target).addEventListener('some-event', pushToEventLog,true);
+
+        ShadyDOM.wrap(target).dispatchEvent(new Event('some-event', {bubbles: true, composed: true}));
+        assert.deepEqual(eventLog, [
+          {currentTarget: outerNoSlot, eventPhase: Event.AT_TARGET},
+          {currentTarget: outerDiv, eventPhase: Event.CAPTURING_PHASE},
+          {currentTarget: innerNoSlot, eventPhase: Event.CAPTURING_PHASE},
+          {currentTarget: target, eventPhase: Event.AT_TARGET},
+          {currentTarget: target, eventPhase: Event.AT_TARGET},
+          {currentTarget: innerNoSlot, eventPhase: Event.BUBBLING_PHASE},
+          {currentTarget: outerDiv, eventPhase: Event.BUBBLING_PHASE},
+          {currentTarget: outerNoSlot, eventPhase: Event.AT_TARGET},
+        ]);
+
+        ShadyDOM.wrap(document.body).removeChild(outerNoSlot);
+      });
+
+      test('events dispatched from children that are disconnected in the ' +
+          'real tree but connected in the user-facing tree bubble up to ' +
+          'parents that are connected in the real tree', () => {
+        customElements.define('parent-listener-element', class extends HTMLElement {
+          constructor() {
+            super();
+
+            // This elements' shadow root is empty, so all of its (non-shadow)
+            // children will be removed from the real tree even though they are
+            // considered connected in the user-facing tree.
+            ShadyDOM.wrapIfNeeded(this).attachShadow({mode: 'open'});
+            ShadyDOM.flush();
+
+            this.listenerCallCount = 0;
+            ShadyDOM.wrap(this).addEventListener('some-event', () => {
+              this.listenerCallCount++;
+            });
+          }
+        });
+
+        customElements.define('child-dispatcher-element', class extends HTMLElement {
+          constructor() {
+            super();
+            ShadyDOM.wrap(this).dispatchEvent(new Event('some-event', {bubbles: true}));
```

Comment:
Using `wrap` instead of `wrapIfNeeded` here was necessary to make this test pass. Should I add `dispatchEvent` to the notes for `wrapIfNeeded`?

https://github.com/webcomponents/polyfills/blob/de5c30c3a51e0890bbecf6dc44b5e8364201730a/packages/shadydom/src/shadydom.js#L83-L93

Reply:
Yes, that sounds good.